File: src/etch/bindings/csharp/transport/FormatFactory.cs
using System;
using System.Collections.Generic;
using System.Text;
using Etch.Msg;
using Etch.Util;

namespace Etch.Transport
{
    public abstract class FormatFactory
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="vf"></param>
        /// <param name="uri"></param>
        /// <returns>the tagged data input with the specified value factory and
        /// initialized by terms from the uri.</returns>
        abstract public TaggedDataInput NewTaggedDataInput( ValueFactory vf, URL uri );

        /// <summary>
        /// 
        /// </summary>
        /// <param name="vf"></param>
        /// <param name="uri"></param>
        /// <returns>the tagged data output with the specified value factory and
        /// initialized by terms from the uri.</returns>
        abstract public TaggedDataOutput NewTaggedDataOutput( ValueFactory vf, URL uri );

        private readonly static Dictionary<String, FormatFactory> formatFactories = new Dictionary<String, FormatFactory>();

        /// <summary>
        /// Gets the named format factory.
        /// </summary>
        /// <param name="name">the name of a configured format factory.</param>
        /// <returns>the named format factory, or null if it isn't defined.</returns>
        public static FormatFactory Get( String name )
        {
            return formatFactories[ name ];
        }

        /// <summary>
        /// Puts a named format factory.
        /// </summary>
        /// <param name="name">the uri scheme of this format factory.</param>
        /// <param name="ff">the format factory to associate with the name.</param>
        public static void Put( String name, FormatFactory ff )
        {
            formatFactories.Add( name, ff );
        }

        static FormatFactory()
        {
            Put( "binary", new BinaryTaggedDataFormatFactory() );
        }
    }
}

File: src/etch/bindings/java/transport/FormatFactory.java
package etch.bindings.java.transport;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

import metreos.util.URL;
import etch.bindings.java.msg.TaggedDataInput;
import etch.bindings.java.msg.TaggedDataOutput;
import etch.bindings.java.msg.ValueFactory;

/**
 * A factory used to construct tagged data processors for messagizer.
 */
abstract public class FormatFactory
{
	/**
	 * @param vf
	 * @param uri
	 * @return the tagged data input with the specified value factory and
	 * initialized by terms from the uri.
	 */
	abstract public TaggedDataInput newTaggedDataInput( ValueFactory vf, URL uri );

	/**
	 * @param vf
	 * @param uri
	 * @return the tagged data output with the specified value factory and
	 * initialized by terms from the uri.
	 */
	abstract public TaggedDataOutput newTaggedDataOutput( ValueFactory vf, URL uri );
	
	private final static Map<String, FormatFactory> formatFactories =
		Collections.synchronizedMap( new HashMap<String, FormatFactory>() );
	
	/**
	 * Gets the named format factory.
	 * @param name the name of a configured format factory.
	 * @return the named format factory, or null if it isn't defined.
	 */
	public static FormatFactory get( String name )
	{
		return formatFactories.get( name );
	}
	
	/**
	 * Puts a named format factory.
	 * @param name the uri scheme of this format factory.
	 * @param ff the format factory to associate with the name.
	 */
	public static void put( String name, FormatFactory ff )
	{
		formatFactories.put( name, ff );
	}
	
	static
	{
		put( "binary", new BinaryTaggedDataFormatFactory() );
	}
}
